pyleros/fedec.py

```diff
@@ -0,0 +1,198 @@
+    combinatorial, except for the updating the pipeline 
+    register. The IM is instantied and only accessed
+    in this stage. The main functions done here are decoding
+    of instruction, setting up branch control signals, selection
+    of other control signals(including the ALU ones), selection 
+    of DM address. an ALU operation on two local variables takes
+    two cycles to execute and another cycle if the result needs 
+    to written back to a local variable
+
+    Arguments (ports):
+        clk: IN Clock signal
+        reset: IN Async reset signal
+        acc: IN Acc value accessed here, not written. This is
+            needed for setting the branch control signals and
+            for memory addredd of JAL
+        dm_data: IN The data read from the DM, which is needed for
+            an direct add or and indirect load/ store(which follows)
+        pipe_dec: OUT List of the decode signals, pass on to the execute stage
+        pipe_imme: OUT Immediate value, as taken from the lower bits 
+                of the instruction, pass on to execute stage
+        pipe_dm_addr: OUT DM read addr, pipeline register
+        pipe_pc: OUT the value of PC, pipeline register
+
+    Parameters:
+
+        filename: Name of the file or a list containing the instructions
+
+    """
+
+    im_addr = Signal(intbv(0)[IM_BITS:])
+
+    instr = Signal(intbv(0)[16:])
+    instr_hi = Signal(intbv(0)[8:])
+
+    branch_en, acc_z = False, True
+
+    pc, pc_next = [Signal(intbv(0)[IM_BITS:])] * 2
+    pc_add = intbv(0)[IM_BITS:]
+
+    decode = [Signal(bool(0)) for i in dlist]
+    decode[int(t_decSignal.op)] = Signal(alu_op_type.LD)
+
+    # Instantiate the instruction memory
+    im_inst = rom.pyleros_im(clk, reset, im_addr, instr, filename=filename)
+
+    # Instantiate the decoder
+    dec_inst = decoder.pyleros_decoder(instr_hi, decode)
+
+    @always_comb
+    def sync_sig():
+
+        # if not reset == reset.active:
+        instr_hi.next = instr[16:8]
+        im_addr.next = pc_next
+        pipe_pc.next = pc_add
+
+
+    @always_comb
+    def dm_addr_sel():
+
+        # if not reset == reset.active:
+        # Note ind ls adds the 8 bit immediate offset
+        # to the previous value of the dm_
+        offset_addr = intbv(dm_data + instr[8:0])[16:]
+        
+        # Indirect Addressing(with offset) 
+        # for indirect load/store
+        if decode[int(t_decSignal.indls)]:
+            pipe_dm_addr.next = offset_addr[DM_BITS:] 
+
+        # Direct Addressing
+        else:
+            pipe_dm_addr.next = instr[DM_BITS:]
+
+    @always_comb
+    def branch_sel():
+
+        # if not reset == reset.active:
+        if acc == 0:
+            acc_z = True
+
+        else:
+            acc_z = False
+
+        branch_en = 0
+
+        if decode[int(t_decSignal.br_op)]:
+            br_type = instr[11:8]
+
+            if br_type == 0b000:
+                # BRANCH
+                branch_en == 1
```

Comment:
If you use interim signals, you can use a combinatorial block to describe how those signals are connected. You don't lose anything at all - the synthesis tool would have to do exactly the same thing with interim variables. So the way I'd design it would be to have a combinatorial (`always_comb`) decoder block, with a synchronous handler block like you're written.